### PR TITLE
Fix AugAssign with templates and literal rhs

### LINKED ISSUE #1427: Pyccelisation of Internal Operation on template variable
### Describe the bug
If we try to use an internal operation on a variable based on a template, pyccel crashed without returning an error message. The Code was working until this commit "Fix AST user tree bug causing problem with AugAssign (https://github.com/pyccel/pyccel/pull/1416)".

### To Reproduce
```python
from pyccel.decorators import template

@template(name='T', types=['float', 'complex'])
def intern_operation_on_template(test_array: 'T'):
    test_array += 1
```

### Error details
```bash
 pyccel kernels.py --language fortran

Traceback (most recent call last):
  File "/home/cato/venv/v_psydac/bin/pyccel", line 33, in <module>
    sys.exit(load_entry_point('pyccel', 'console_scripts', 'pyccel')())
  File "/home/cato/Code/pyccel/pyccel/commands/console.py", line 239, in pyccel
    execute_pyccel(filename,
  File "/home/cato/Code/pyccel/pyccel/codegen/pipeline.py", line 239, in execute_pyccel
    fname, prog_name = codegen.export(fname, language=language)
  File "/home/cato/Code/pyccel/pyccel/codegen/codegen.py", line 205, in export
    code = self._printer.doprint(self.ast)
  File "/home/cato/Code/pyccel/pyccel/codegen/printing/codeprinter.py", line 52, in doprint
    lines = self._print(expr).splitlines(True)
  File "/home/cato/Code/pyccel/pyccel/codegen/printing/codeprinter.py", line 88, in _print
    obj = getattr(self, print_method)(expr)
  File "/home/cato/Code/pyccel/pyccel/codegen/printing/fcode.py", line 466, in _print_Module
    func_strings = [''.join([sep, self._print(i), sep]) for i in expr.funcs]
  File "/home/cato/Code/pyccel/pyccel/codegen/printing/fcode.py", line 466, in <listcomp>
    func_strings = [''.join([sep, self._print(i), sep]) for i in expr.funcs]
  File "/home/cato/Code/pyccel/pyccel/codegen/printing/codeprinter.py", line 88, in _print
    obj = getattr(self, print_method)(expr)
  File "/home/cato/Code/pyccel/pyccel/codegen/printing/fcode.py", line 1929, in _print_FunctionDef
    body_code = self._print(expr.body)
  File "/home/cato/Code/pyccel/pyccel/codegen/printing/codeprinter.py", line 88, in _print
    obj = getattr(self, print_method)(expr)
  File "/home/cato/Code/pyccel/pyccel/codegen/printing/fcode.py", line 1524, in _print_CodeBlock
    line = self._print(b)
  File "/home/cato/Code/pyccel/pyccel/codegen/printing/codeprinter.py", line 88, in _print
    obj = getattr(self, print_method)(expr)
  File "/home/cato/Code/pyccel/pyccel/codegen/printing/fcode.py", line 2069, in _print_AugAssign
    expr.invalidate_node()
  File "/home/cato/Code/pyccel/pyccel/ast/basic.py", line 90, in invalidate_node
    c.remove_user_node(self)
  File "/home/cato/Code/pyccel/pyccel/ast/basic.py", line 384, in remove_user_node
    assert(user_node in self._user_nodes)

```

### PR DESCRIPTION
**Pull request by EmilyBourne**
Add the creation stage to all Pyccel AST objects and use this property to avoid deleting relevant objects from the AST tree when reusing basic objects between the syntactic and semantic stage. Fixes #1427

**Modifications list:**

- Add property `pyccel_staging` and method `update_pyccel_staging` to class `Basic`;
- Replace method `clear_user_nodes` with new method `clear_syntactic_user_nodes` in class `Basic`;
- Add property `current_stage` to class `PyccelStage`;
- In semantic parser replace most calls to `clear_user_nodes()` with successive calls to `clear_syntactic_user_nodes()` and `update_pyccel_staging()`;
- Add unit test to prove that issue #1427 is fixed.

### CODE REVIEW

**Comment by EmilyBourne** [comment]
The coverage is failing as the following functions are unused:
- `_visit_MacroShape` - See #1372 
- `_visit_AnnotatedComment` - This is partially implemented for OpenACC. As it is not fully implemented there are no tests. This code could be removed to ensure a neat error is raised, or left as its contents are simply the same two lines which appear in other functions

File: tests/epyccel/modules/generic_functions_2.py [hunk 1]
```diff
@@ -137,6 +137,11 @@ def mix_types_3(x, y):
     return -x
 
 
+@template('T', types=['int64[:]', 'real[:]', 'complex[:]'])
+@types('T')
+def mix_array_scalar(x):
+    x[:] += 1
+
 @template('T', types=['int64[:]', 'real[:]', 'complex[:]'])
```
Comment (yguclu) [inline tests/epyccel/modules/generic_functions_2.py:145]: I thought that `real` was deprecated in favor of `float`... :stuck_out_tongue_closed_eyes:

Comment (EmilyBourne) [reply tests/epyccel/modules/generic_functions_2.py]: Fixed for this file : 5b2dc7fa

**Review by yguclu — APPROVED** [review]
Thanks @EmilyBourne! This looks good to me

**Pull request merged**